### PR TITLE
Limit chatting to users in your group and allow to exclude certain groups

### PR DESCRIPTION
**Pull request by LEDfan**
This PR integrates the changes provided in https://github.com/nextcloud/server/pull/5585 so an admin can limit Chatting in the following two ways, using the Sharing settings in core:
 - Exclude certain groups from chatting:
afbeelding
 - Restrict sharing to own groups:
afbeelding

This needs the upstream PR + another (not yet made) PR to make the ContactsStore a public API in Nextcloud, therefore the tests should and will fail on stable12 and master.  (https://gist.github.com/LEDfan/6baa61a6d23ed5dff2fac877bc9e2895)

TODO: 
 - [x] test `Exclude certain groups from chatting` and make sure the roster isn't loaded for this user
 - [x] more manually testing
 - [x] tests
   -  [x] removeRosterItemForUsersInGroup
   - [x] MessageHandler illegal message
   - [x] getConnectedUsers with different sharing settings
   - [x] roster with different sharing settings
 - [x]  exclude user from roster if that user can't talk to the current user
 - [x] when adding a user to a group, send a roster push for all users in that group
 - [x] when removing a user from a group, send a roster push for all users in that group
 - [x] test if it the app still works on stable10, stable11 and stable12
 - [x] Upstream PR: https://github.com/nextcloud/server/pull/5585
 - [x] Upstream PR: https://github.com/jsxc/jsxc/pull/629
 - [x] Upstream PR: https://github.com/nextcloud/server/pull/6637

Fixes jsxc/jsxc#306

### CODE REVIEW

File: lib/hooks.php [hunk 1]
```diff
@@ -112,4 +128,15 @@ public function onChangeUser(IUser $user, $feature, $value)
...
 	}
+
+	public function onAddUserToGroup(IGroup $group, IUser $user)
+	{
+		$this->rosterPush->createOrUpdateRosterItem($user);
+	}
+
+	public function onRemoveUserFromGroup(IGroup $group, IUser $user)
+	{
+		$this->rosterPush->removeRosterItemForUsersInGroup($group, $user->getUID());
```
Comment (LEDfan) [inline lib/hooks.php]: TODO this must only be done when one of the privacy options is enabled and when running on version > 12.0.3

**Comment by LEDfan** [comment]
@sualko FYI this is something not for 3.3, it needs to much changes upstream.

**Comment by sualko** [comment]
Thanks for the information :+1:

**Comment by Schmuuu** [comment]
Hi,

I pretty urgent need the feature "Restrict sharing to own groups". Is there an eta when this feature will be released or in which version this will come?
I'm no coder unfortunately but can I provide some help in a different way? I'm willing to test of course if you need help there.

And thanks for the very good work so far with this great app!

**Comment by LEDfan** [comment]
@Schmuuu I'm currently having exams again, I'll finish this PR in approx 3 weeks.

**Comment by MarcelWaldvogel** [comment]
@Schmuuu Should only a limited group have access to the chat feature, or should they only be able to communicate within the group?

The latter can be achieved by setting up an external XMPP server and creating appropriate rules.

However, for both cases I do not see a real reason to have this; users can communicate anyway, outside of JSXC (and possible with more side-effects and information leakage).

**Comment by Schmuuu** [comment]
Hi @MarcelWaldvogel 

> or should they only be able to communicate within the group?
Yes, that is what I would like to have.

I configured the sharing with:
"Restrict users to only share with users in their groups"

It would just be consistent if I can either configure the chat manually or if the app uses the sharing setting automatically and restricts users to chat with other users in their group only. 

I have the following situation:
+ the users in different groups don't know each other and don't want to know each other
+ in return users in different groups don't want to chat with each other 
+ users in different groups can't share files anyway (and don't want to)

What I would like to achieve:
+ provide the users a clean user list (in chat bar) of online/ offline users (not a list of 30+ users where 80% of the listed users is not of any interest)
+ don't want to show everyone who is using the server
+ create a test user for friends to show them the nice features of nextcloud and not(!) automatically show them the full user list 
+ show the chat feature to that test user and allow these test users to chat with me (because my user is in that test group as well)

I hope you agree, that there are just some use cases, where it is just perfect to have the chance to limit chats within a group only.

While I'm trying to convince my company to use a full-featured Nextcloud, this would be critical as well. Providing customers the chat feature would be nice, but we would need to hide all the customers from each other, meaning only the users of one customer (one company) can chat with us and each other, but not the users of different customers with each other.

**Pull request merged**